[GeoMechanicsApplication] Added a Python script for running a settlement calculation

avdg81 opened the pull request:
**📝 Description**
This PR adds a Python script that runs a settlement calculation using the same DLL entry point as D-Settlement. The aim of this script is to easily reproduce issues reported for D-Settlement.

avdg81 committed:
Added a Python script for running the settlement workflow

This is a rudimentary version that needs to be refined and cleaned up.

avdg81 committed:
Improved error message and return code

When running an analysis stage fails, print and return the exit code of `RunStage`.

avdg81 committed:
Formatted the Python code using `black`

avdg81 committed:
Updated the help text of a command line argument

avdg81 commented on applications/GeoMechanicsApplication/python_scripts/run_geo_settlement.py:
```diff
@@ -0,0 +1,53 @@
+
+    args = parser.parse_args()
+
+    import KratosMultiphysics.GeoMechanicsApplication as GeoMechanicsApplication
```
I haven't put this `import` statement at the top of the file, since doing that would result in always printing the loaded applications, even when asking for the help message (i.e. by running `python run_geo_settlement.py -h`).

avdg81 commented on applications/GeoMechanicsApplication/python_scripts/run_geo_settlement.py:
The file has been formatted using `black`.

WPK4FEM commented on applications/GeoMechanicsApplication/python_scripts/run_geo_settlement.py:
```diff
@@ -0,0 +1,53 @@
+        "-w", "--working-dir", help="working directory of the analysis run", default="."
+    )
+    parser.add_argument(
+        "project_parameters_file",
```
Files for multiple stages would be more than 1, so I would expect _files i.s.o. _file.

avdg81 replied:
I have solved the problem by passing a value for `metavar` to `add_argument`. When retrieving the list of filenames, we now see the plural form (as you requested). In the help text, however, we see the singular form, since it presents the argument as a list of at least one filename: `usage: run_geo_settlement.py [-h] [-w WORKING_DIR] PROJECT_PARAMETERS_FILENAME [PROJECT_PARAMETERS_FILENAME ...]`.

WPK4FEM reviewed (dismissed):
No blocking comments from my side, I've seen it demonstrated at Anne's system.

markelov208 commented on applications/GeoMechanicsApplication/python_scripts/run_geo_settlement.py:
```diff
@@ -0,0 +1,53 @@
+
+
+def _main() -> int:
+    parser = argparse.ArgumentParser()
```
Would it be nicer to include a title like that?

parser = argparse.ArgumentParser(description="Run GeoMechanics settlement analysis.")

avdg81 replied:
Yes, that is a nice suggestion. It will help the user to better understand the intention of the script. Done.

markelov208 commented on applications/GeoMechanicsApplication/python_scripts/run_geo_settlement.py:
```diff
@@ -0,0 +1,53 @@
+    )
+    atexit.register(clean_up_api_object, settlement_api)
+
+    for project_parameters_filename in args.project_parameters_file:
```
Should a try block be added, for example, to catch problems with numerics?

avdg81 replied:
This shouldn't be needed, since `RunStage` already catches all exceptions. When an exception is caught, it will return a nonzero exit code.

markelov208 reviewed (dismissed):
Hi Anne, thank you for the script. Hope I will use it. I have two non-blocking comments.

avdg81 committed:
Merge remote-tracking branch 'origin/master' into geo/settlement-app

avdg81 committed:
Processed the review suggestions

- Make a better distinction between singular and plural forms of filename and filenames.
- The usage message now includes a description of what the script's aim is.

markelov208 reviewed (approved):
Hi Anne, thank you for the clarification. I think this PR is good to merge.
